hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/TestCGroupsV2HandlerImpl.java

```diff
@@ -0,0 +1,270 @@
+   */
+  @Test
+  public void testMtabParsing() throws Exception {
+    // Initialize mtab and cgroup dir
+    File parentDir = new File(tmpPath);
+    // create mock cgroup
+    File mockMtabFile = createPremountedCgroups(parentDir);
+
+    CGroupsV2HandlerImpl cGroupsHandler = new CGroupsV2HandlerImpl(
+        createMountConfiguration(),
+        privilegedOperationExecutorMock, mockMtabFile.getAbsolutePath());
+
+    // Run mtabs parsing
+    Map<String, Set<String>> newMtab =
+            cGroupsHandler.parseMtab(mockMtabFile.getAbsolutePath());
+    Map<CGroupsHandler.CGroupController, String> controllerPaths =
+            cGroupsHandler.initializeControllerPathsFromMtab(
+            newMtab);
+
+    // Verify
+    Assert.assertEquals(4, controllerPaths.size());
+    assertTrue(controllerPaths
+        .containsKey(CGroupsHandler.CGroupController.CPU));
+    assertTrue(controllerPaths
+        .containsKey(CGroupsHandler.CGroupController.MEMORY));
+    String cpuDir = controllerPaths.get(CGroupsHandler.CGroupController.CPU);
+    String memoryDir =
+        controllerPaths.get(CGroupsHandler.CGroupController.MEMORY);
+    Assert.assertEquals(parentDir.getAbsolutePath(), cpuDir);
+    Assert.assertEquals(parentDir.getAbsolutePath(), memoryDir);
+  }
+
+  @Test
+  public void testManualCgroupSetting() throws Exception {
+    YarnConfiguration conf = new YarnConfiguration();
+    conf.set(YarnConfiguration.NM_LINUX_CONTAINER_CGROUPS_MOUNT_PATH, tmpPath);
+    conf.set(YarnConfiguration.NM_LINUX_CONTAINER_CGROUPS_HIERARCHY,
+        "/hadoop-yarn");
+
+    File subCgroup = new File(tmpPath, "/hadoop-yarn");
+    Assert.assertTrue("temp dir should be created", subCgroup.mkdirs());
+    subCgroup.deleteOnExit();
+
+    String enabledControllers = "cpuset cpu io memory hugetlb pids rdma misc\n";
+    createFileWithContent(subCgroup, CGroupsHandler.CGROUP_CONTROLLERS_FILE, enabledControllers);
+
+    File subtreeControlFile = new File(subCgroup.getAbsolutePath(),
+        CGroupsHandler.CGROUP_SUBTREE_CONTROL_FILE);
+    Assert.assertTrue("empty subtree_control file should be created",
+        subtreeControlFile.createNewFile());
+
+    CGroupsV2HandlerImpl cGroupsHandler = new CGroupsV2HandlerImpl(conf, null);
+    cGroupsHandler.initializeCGroupController(CGroupsHandler.CGroupController.CPU);
+
+    Assert.assertEquals("CPU cgroup path was not set", subCgroup.getAbsolutePath(),
+            new File(cGroupsHandler.getPathForCGroup(
+                CGroupsHandler.CGroupController.CPU, "")).getAbsolutePath());
+
+    // Verify that the subtree control file was updated
+    String subtreeControllersEnabledString = FileUtils.readFileToString(subtreeControlFile,
+        StandardCharsets.UTF_8);
+    Set<String> subtreeControllersEnabled = new HashSet<>(Arrays.asList(
+        subtreeControllersEnabledString.replace("+", "").split(" ")));
+
+    Assert.assertEquals("The newly added controllers doesn't contain + signs",
```

Comment:
For easier comprehension, I would add this assert next to the readFileToString (at 240) so it is easier to see that we work on the file content that we just read.

Reply:
Done.